### PR TITLE
OLED-427 add prom metrics server

### PR DESCRIPTION
**Pull request by jaskerv**
*Issue #, if available:*

*Description of changes:*
- Refactor how application and host logs are identified.
- Add custom metrics using otel:
  - Output record total `fluentbit_output_record_total` - The count of records, this has a dimension for `service_name`
  - Output dropped body total `fluentbit_output_dropped_body_total` - The count of records that do not contain log field or message field.
 

Note:
- All `.` are replaced with `_` with prometheus.
- Byte size is difficult to calculate due to the use of `map[interface{}]interface{}` this is difficult to convert. I will investigate this in another PR.
- Incase metric fails to infer service, `_missing_service_name` will be used instead.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: enricher/eks/eks.go
Comment (sasanrose) [inline enricher/eks/eks.go:15]: nit: we can change this to an enum type

File: enricher/eks/metric.go
Comment (sasanrose) [inline enricher/eks/metric.go]: Are we going to add a context in the future to `AddRecordCount`? if not we can change TODO to Background

File: enricher/eks/metric.go
Comment (sasanrose) [inline enricher/eks/metric.go]: nit: Can't we just change the type cast here to `map[interface{}]string` or even `map[string]string`? So we can set `serviceName` to string and avoid the `fmt.Sprintf` at the end. Same for resources

Comment (sasanrose) [reply enricher/eks/metric.go]: Also can we change this to the same way we are checking for `RESOURCE_FIELD_NAME`

File: metricserver/metricserver.go
Comment (sasanrose) [inline metricserver/metricserver.go:30]: Do we need to implement a shutdown function to gracefully shutdown the metric server in the `fluent-bit-kinesis.go`? Seems that the MetricProvider supports this: https://pkg.go.dev/go.opentelemetry.io/otel/sdk/metric#MeterProvider.Shutdown

**Comment by sasanrose** [comment]
Great job @Jaskerv. LGTM in general just left a couple of comments

File: enricher/eks/metric.go
Comment (jaskerv) [reply enricher/eks/metric.go]: Regarding `map[interface{}]string`, I encountered an issue in deployment `telemetry-test` where it fails to cast `interface{}` to `string`.

```
panic: interface conversion: interface {} is []uint8, not string [recovered]
        panic: interface conversion: interface {} is []uint8, not string
```

This is why I have to do it as `fmt.Sprintf` but let me know if there a better way to tackle this.

Sidenote - I want to also say that this `map[interface{}]interface{}` is really painful to deal with 😢

File: enricher/eks/metric.go
Comment (jaskerv) [reply enricher/eks/metric.go]: AFAIK, there is no intention but I could see the complexity of this growing so additional telemetry could be used eg the use of traces etc. But for now I will change this to background.

File: enricher/eks/metric.go
Comment (mallyx3) [inline enricher/eks/metric.go]: Lets use mappings.PLACEHOLDER_MISSING_KUBERNETES_METADATA for consistency with the rest of our pipeline

File: metricserver/metricserver.go
Comment (jaskerv) [reply metricserver/metricserver.go:30]: I thought about this too. 

From the description of `Shutdown` it looks like it just flushes and releases any computational resources.

For our implementation, we use a pull-based approach using a Prometheus exporter. With this, we aren't buffering any metrics to send to an external agent.

IIUC, if the main thread gets terminated, MetricProvider and Prometheus Server should be killed along with it, which I may be wrong but would also release computational resources? 

WDYT?
Happy to implement a shutdown procedure for MeterProvider.

File: enricher/eks/metric.go
Comment (sasanrose) [reply enricher/eks/metric.go]: Oh ok so then you can probably use type `rune` and then use `string()` to convert it. I think it still would be nicer than `fmt.Sprintf` 😄

File: metricserver/metricserver.go
Comment (sasanrose) [reply metricserver/metricserver.go:30]: hmm you're right. in that case we may want to have a graceful shutdown just for the http server. Meaning if the `/metric` endpoint is being called and we receive a shutdown signal while we are pulling , it will wait for the http server to gracefully shutdown. However, not sure how important it is to not loosing these metrics. Maybe it does not worth the effort and we can afford to loose them

**Comment by jaskerv** [comment]
Hey team, there might be an issue with how the Datadog agent scrapes our fluent-bit agent. If the approach changes, I would need to update this PR to reflect that. I'll put this PR on hold for now until we have an approach for that.

Thanks!

File: metricserver/metricserver.go
Comment (jaskerv) [reply metricserver/metricserver.go:30]: IIUC, there is no blocking thread as `main()` is not executed. I believe the go functions are called as C bindings so doing any grace shutdown would be tricky. WDYT?

**Comment by jaskerv** [comment]
I managed to get service discovery working with 2 instances. See [here](https://app.datadoghq.com/metric/explorer?start=1687147609653&end=1687149409653&paused=false#N4Ig7glgJg5gpgFxALlAGwIYE8D2BXJVEADxQEYAaELcqyKBAC1pEbghkcLIF8qo4AMwgA7CAgg4RKUAiwAHOChASAtnADOcAE4RNIKtrgBHPJoQaUAbVBGN8qVoD6gnNtUZCKiOq279VKY6epbINiAiGOrKQdpYZAYgUJ4YThr42gDGSsgg6gi6mZaBZnHKGniqyBhQUI4AdIJoZiIIAEbiTvgI8gRO8to4mU5GmW5QGvVjeK3Amc0aCDpOkerIS2hw+XEAtEuLPPUYGk7TrQAUAJQgfKCrOSCxWABMickIqel4WQ-bEEWJJ7lSrVWoNJotdqdbq9BAjOBjbRQKb4WbzPCLZb3dZwTbbLB7cw8AAEbSwxOAfgAbv84CsonBDsdTqiEFcbhQ7gyYqUsABmN4pNIZbLKP4Ako6Gi5CpVGp1ESTCFwVodOEwvpQQbyRRQJxtHBQLAomYIOYLJbaelrDZbRC7fYIJknM5s648AC6VFc7jwmFC4R9qj9GB5UoStxAQZDYbir0j0cwsf5Ny9IEWWE2MhA8gwmwQS2UdRgp39Gn+iTQojpckU5RwVaglerTnoTGUIjcHjQHKSEHsmCwTlrDyrIiUnp4fHT8irCAAwlJhDAUCI-WgeEA) to most recent deployment and the custom metrics. We can go ahead with this PR.

File: enricher/eks/metric.go
Comment (jaskerv) [reply enricher/eks/metric.go]: Do you mean like this?
```
func inferServiceName(record map[interface{}]interface{}, logType LogType) string {
	// fallback in case unable to find service name.
	var serviceName []rune = []rune(mappings.PLACEHOLDER_MISSING_KUBERNETES_METADATA)

	k8sPayload, ok := record[mappings.KUBERNETES_RESOURCE_FIELD_NAME].(map[interface{}]interface{})
	// k8s field exists, set default to container name.
	if ok {
		serviceName = k8sPayload[mappings.KUBERNETES_CONTAINER_NAME].([]rune)
	}

	switch logType {
	case TYPE_APPLICATION:
		if labels, ok := k8sPayload[mappings.KUBERNETES_LABELS_FIELD_NAME].(map[interface{}][]rune); ok {
			if val, ok := labels[mappings.KUBERNETES_LABELS_NAME]; ok {
				serviceName = val
			}
		}
	case TYPE_HOST:
		if resource, ok := record[mappings.RESOURCE_FIELD_NAME].(map[interface{}][]rune); ok {
			if val, ok := resource[mappings.RESOURCE_SERVICE_NAME]; ok {
				serviceName = val
			}
		}
	}

	return string(serviceName)
}
```
I think this is a good approach but it's tricky because I believe the incoming data is in bytes which I have to convert to string anyways. With my current tests, I get a conversion error when trying strings or []byte in test.

WDYT?

Comment (sasanrose) [reply enricher/eks/metric.go]: yeah it seems either way is confusing so we may as well leave as it is

File: metricserver/metricserver.go
Comment (sasanrose) [reply metricserver/metricserver.go:30]: It seems that Go already has taken care of the complexities:
https://pkg.go.dev/os/signal#hdr-Non_Go_programs_that_call_Go_code

Comment (Jaskerv) [reply metricserver/metricserver.go:29]: I just pushed a shutdown function and it would be invoked in `FLBPluginExit`, this function is exported and will be called when shutdown occurs. This implementation isn't the cleanest approach.

Testing this in telemetry-test when killing the pod gives:
```
time="2023-06-20T03:44:19Z" level=info msg="shutting down metrics server"
time="2023-06-20T03:44:19Z" level=fatal msg="http: Server closed"
```

Comment (jaskerv) [reply metricserver/metricserver.go:30]: I added a Shutdown procedure in `FLBPluginExit` which is called when exiting. I changed how the Server starts.

The setup isn't ideal and clean but it works. WDYT?

```
time="2023-06-20T03:54:34Z" level=info msg="shutting down metrics server"
time="2023-06-20T03:54:34Z" level=info msg="http: Server closed"
time="2023-06-20T03:54:34Z" level=info msg="shutdown metrics server complete"
```

File: fluent-bit-kinesis.go
Comment (sasanrose) [inline fluent-bit-kinesis.go]: nit: Do we want the server shutdown to return an error so we can return `output.FLB_ERROR` instead of `FLB_OK`?

File: metricserver/metricserver.go
Comment (sasanrose) [reply metricserver/metricserver.go:30]: yeap, makes sense to leave it to FluentBit to handle the signal and close the plugins

**Review by sasanrose — APPROVED** [review]
LGTM! just left a nit comment on the shutdown

File: enricher/eks/metric.go
Comment (jaskerv) [reply enricher/eks/metric.go]: @mallyx3 I had some more things on this, I think we should use another label, the reason is if it fails to find service name, this shouldn't pollute `PLACEHOLDER_MISSING_KUBERNETES_METADATA`. WDYT?

Comment (jaskerv) [reply enricher/eks/metric.go]: Discussed offline. Did not want to pollute `PLACEHOLDER_MISSING_KUBERNETES_METADATA` so defaulting to `_missing_service_name`

File: enricher/eks/metric.go
Comment (mallyx3) [inline enricher/eks/metric.go:22]: What is this line doing?

File: enricher/eks/metric.go
Comment (mallyx3) [inline enricher/eks/metric.go:52]: Where can i read about the decision to name the attribute to be `source` ?

**Review by mallyx3 — APPROVED** [review]

File: enricher/eks/metric.go
Comment (jaskerv) [reply enricher/eks/metric.go:22]: This sets the scope of the meter. [Reference](https://opentelemetry.io/docs/specs/otel/metrics/api/#get-a-meter)

**Pull request merged**
